src/providers/mdal/qgsmdalprovider.cpp

```diff
@@ -120,56 +118,83 @@ bool QgsMdalProvider::addDataset( const QString &uri )
   }
 }
 
-int QgsMdalProvider::datasetCount() const
+int QgsMdalProvider::datasetGroupCount() const
 {
-  return MDAL_M_datasetCount( mMeshH );
+  return MDAL_M_datasetGroupCount( mMeshH );
 }
 
-QgsMeshDatasetMetadata QgsMdalProvider::datasetMetadata( int datasetIndex ) const
+
+int QgsMdalProvider::datasetCount( int groupIndex ) const
 {
-  if ( datasetIndex >= mDatasets.length() )
-    return QgsMeshDatasetMetadata();
+  DatasetGroupH group = MDAL_M_datasetGroup( mMeshH, groupIndex );
+  if ( !group )
+    return 0;
+  return MDAL_G_datasetCount( group );
+}
 
-  if ( datasetIndex < 0 )
-    return QgsMeshDatasetMetadata();
+QgsMeshDatasetGroupMetadata QgsMdalProvider::datasetGroupMetadata( int groupIndex ) const
+{
+  DatasetGroupH group = MDAL_M_datasetGroup( mMeshH, groupIndex );
+  if ( !group )
+    return QgsMeshDatasetGroupMetadata();
 
-  DatasetH dataset = mDatasets[datasetIndex];
 
-  bool isScalar = MDAL_D_hasScalarData( dataset );
-  bool isValid = MDAL_D_isValid( dataset );
-  bool isOnVertices = MDAL_D_isOnVertices( dataset );
+  bool isScalar = MDAL_G_hasScalarData( group );
+  bool isOnVertices = MDAL_G_isOnVertices( group );
+  QString name = MDAL_G_name( group );
 
   QMap<QString, QString> metadata;
-  int n = MDAL_D_metadataCount( dataset );
+  int n = MDAL_G_metadataCount( group );
   for ( int i = 0; i < n; ++i )
   {
-    QString key = MDAL_D_metadataKey( dataset, i );
-    QString value = MDAL_D_metadataValue( dataset, i );
+    QString key = MDAL_G_metadataKey( group, i );
+    QString value = MDAL_G_metadataValue( group, i );
     metadata[key] = value;
   }
 
-  QgsMeshDatasetMetadata meta(
+  QgsMeshDatasetGroupMetadata meta(
+    name,
     isScalar,
-    isValid,
     isOnVertices,
     metadata
   );
 
   return meta;
 }
 
-QgsMeshDatasetValue QgsMdalProvider::datasetValue( int datasetIndex, int valueIndex ) const
+QgsMeshDatasetMetadata QgsMdalProvider::datasetMetadata( QgsMeshDatasetIndex index ) const
 {
-  if ( datasetIndex >= mDatasets.length() )
+  DatasetGroupH group = MDAL_M_datasetGroup( mMeshH, index.group() );
+  if ( !group )
+    return QgsMeshDatasetMetadata();
+
+  DatasetH dataset = MDAL_G_dataset( group, index.dataset() );
```

Comment:
Also return earlier if the returned dataset is invalid?

Reply:
done